### PR TITLE
Added AMI and ARI scoring strategies

### PR DESCRIPTION
**Pull request by ajsedgewick-nh**
Adds scores based on Adjusted Rand Index and Adjusted Mutual Information. Currently uses the intersection of the sets of alarms used in the two situation lists and will return 0 if either list is empty.

### CODE REVIEW

File: features/score/src/main/java/org/opennms/alec/features/score/impl/AMIScoringStrategy.java [hunk 1]
```diff
@@ -0,0 +1,159 @@
+/*******************************************************************************
+ * This file is part of OpenNMS(R).
+ *
+ * Copyright (C) 2018 The OpenNMS Group, Inc.
```
Comment (j-white) [inline features/score/src/main/java/org/opennms/alec/features/score/impl/AMIScoringStrategy.java]: Update 2018 to the current year.

File: features/score/src/main/java/org/opennms/alec/features/score/impl/AMIScoringStrategy.java [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+ * OpenNMS(R) is Copyright (C) 1999-2018 The OpenNMS Group, Inc.
```
Comment (j-white) [inline features/score/src/main/java/org/opennms/alec/features/score/impl/AMIScoringStrategy.java]: Similarly here.

File: features/score/src/main/java/org/opennms/alec/features/score/impl/AMIScoringStrategy.java [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+
+import org.opennms.alec.datasource.api.Alarm;
+import org.opennms.alec.datasource.api.Situation;
+import org.opennms.alec.features.score.api.ScoreMetric;
+import org.opennms.alec.features.score.api.ScoreReport;
+import org.opennms.alec.features.score.api.ScoringStrategy;
+
+import com.google.common.collect.Maps;
+import org.apache.commons.math3.special.Gamma;
+//import org.apache.commons.math3.util.CombinatoricsUtils;
```
Comment (j-white) [inline features/score/src/main/java/org/opennms/alec/features/score/impl/AMIScoringStrategy.java]: Remove unused imports.

File: features/score/src/main/java/org/opennms/alec/features/score/impl/AMIScoringStrategy.java [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+
+        double ami = 0.0;
+        if(alarmIds.size()>0){
+            ami = (mutualInfo - emi) / ((.5*(hu+hv)) - emi);
+        }
+
+
+        final ScoreReport scoreReport = new ScoreReport();
+        scoreReport.setScore(ami);
+        scoreReport.getMetrics().add(new ScoreMetric("rawMutualInformation", mutualInfo, "TODO"));
```
Comment (j-white) [inline features/score/src/main/java/org/opennms/alec/features/score/impl/AMIScoringStrategy.java]: Can we replace the TODOs with a basic textual description?

File: features/score/src/main/java/org/opennms/alec/features/score/impl/ARIScoringStrategy.java [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+
+        // From Hubert and Arabie (1985) see https://jmlr.csail.mit.edu/papers/volume11/vinh10a/vinh10a.pdf
+        double score = 0.0;
+        if(alarmIds.size()>0){
+            score = (2.0 * ((tp*tn) - (fp*fn)))/(((tp+fn)*(fn+tn)) + ((tp+fp)*(fp+tn)));
+        }
+
+        final ScoreReport scoreReport = new ScoreReport();
+        scoreReport.setScore(score);
+        scoreReport.getMetrics().add(new ScoreMetric("TP", tp, "TODO"));
```
Comment (j-white) [inline features/score/src/main/java/org/opennms/alec/features/score/impl/ARIScoringStrategy.java]: Similarly here.

File: features/score/src/test/java/org/opennms/alec/features/score/impl/ScoringStrategyTest.java [hunk 6]
```diff
@@ -76,13 +78,16 @@ public void canComputeScores() {
...
 
-        // Comparing a situation with a single alarm to an empty situation should generate a score greater than 0
-        Situation situation = ImmutableSituation.newBuilderNow()
-                .setId("test")
-                .addAlarm(ImmutableAlarm.newBuilder().setId("test").build())
-                .build();
-        report = scorer.score(Sets.newHashSet(situation), Sets.newHashSet(emtpySituation));
-        assertThat(report.getScore(), Matchers.greaterThan(0.0d));
+        // ARI and AMI won't pass this test unless we add missing alarms as singletons
+        if(!scorer.getName().equals("ari") && !scorer.getName().equals("ami")){
```
Comment (j-white) [inline features/score/src/test/java/org/opennms/alec/features/score/impl/ScoringStrategyTest.java]: Is the expected score 0 for these then?

File: features/score/src/test/java/org/opennms/alec/features/score/impl/ScoringStrategyTest.java [hunk 7]
```diff
@@ -76,13 +78,16 @@ public void canComputeScores() {
...
+            // Comparing a situation with a single alarm to an empty situation should generate a score greater than 0
+            Situation situation = ImmutableSituation.newBuilderNow()
+                    .setId("test")
+                    .addAlarm(ImmutableAlarm.newBuilder().setId("test").build())
+                    .build();
+            report = scorer.score(Sets.newHashSet(situation), Sets.newHashSet(emtpySituation));
+            assertThat(report.getScore(), Matchers.greaterThan(0.0d));
```
Comment (j-white) [inline features/score/src/test/java/org/opennms/alec/features/score/impl/ScoringStrategyTest.java:93]: I'd consider adding the `if` block around the assertion and testing for 0 in this case.

**Review by j-white — COMMENTED** [review]
Thanks for putting this together @ajsedgewick-nh.  Mostly minor nits.

File: features/score/src/test/java/org/opennms/alec/features/score/impl/ScoringStrategyTest.java (hunk 7 above)
Comment (ajsedgewick-nh) [reply features/score/src/test/java/org/opennms/alec/features/score/impl/ScoringStrategyTest.java:93]: yup 0 is the expected score here so this sounds good for now. we may want to discuss how cluster engines and scores deal with singletons in the future.

**Comment by j-white** [comment]
Thanks for the update - changes look good 👍

**Pull request merged**